File: Parciales/septiembre2020/src/AVL.java
/**
 * Student's name:
 *
 * Student's group:
 */

import dataStructures.list.List;
import dataStructures.list.LinkedList;

import java.util.Iterator;


class Bin {
    private int remainingCapacity; // capacity left for this bin
    private List<Integer> weights; // weights of objects included in this bin

    public Bin(int initialCapacity) {
        // todo
        this.remainingCapacity = initialCapacity;
    }

    // returns capacity left for this bin
    public int remainingCapacity() {
        // todo
        return remainingCapacity;
    }

    // adds a new object to this bin
    public void addObject(int weight) {
        // todo
        if(remainingCapacity - weight >= 0){
            this.weights.append(weight);
        } else {
            throw new RuntimeException("El objeto no cabe en el cubo");
        }
    }

    // returns an iterable through weights of objects included in this bin
    public Iterable<Integer> objects() {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        sb.append(remainingCapacity);
        sb.append(", ");
        sb.append(weights.toString());
        sb.append(")");
        return sb.toString();
    }
}

// Class for representing an AVL tree of bins
public class AVL {
    static private class Node {
        Bin bin; // Bin stored in this node
        int height; // height of this node in AVL tree
        int maxRemainingCapacity; // max capacity left among all bins in tree rooted at this node
        Node left, right; // left and right children of this node in AVL tree

        // recomputes height of this node
        void setHeight() {
            // todo
            height = Math.max(left.height, right.height) + 1;
        }

        // recomputes max capacity among bins in tree rooted at this node
        void setMaxRemainingCapacity() {
            // todo
            maxRemainingCapacity = Math.max(left.maxRemainingCapacity, right.maxRemainingCapacity);
        }

        // left-rotates this node. Returns root of resulting rotated tree
        Node rotateLeft() {
            // todo
            if(right != null) {
                Node root = null;

                root.bin = this.right.bin;
                root.left.bin = this.bin;
                root.left.left = this.left;
                root.left.right = this.right.left;
                root.right = this.right.right;
                root.height = right.height;
                root.left.height = height;

                return root;
            }
            return this;
        }
    }

    private static int height(Node node) {
        // todo
        return node.height;
    }

    private static int maxRemainingCapacity(Node node) {
        // todo
        return node.maxRemainingCapacity;
    }

    private Node root; // root of AVL tree

    public AVL() {
        this.root = null;
    }

    // adds a new bin at the end of right spine.
    private void addNewBin(Bin bin) {
        // todo
        root = addNewBinRec(bin, root);
    }

    private Node addNewBinRec(Bin bin, Node node){
        if(node.right == null){
            node.right.bin = bin;
            node.right.setHeight();
        } else {
            node = addNewBinRec(bin, node.right);
        }

        if(node.right.height > node.left.height + 1){
            node.rotateLeft();
        }

        return node;
    }

    // adds an object to first suitable bin. Adds
    // a new bin if object cannot be inserted in any existing bin
    public void addFirst(int initialCapacity, int weight) {
        // todo
        root = addFirstRec(initialCapacity, weight, root);
    }

    public Node addFirstRec(int initialCapacity, int weight, Node node){
        if(node == null) {
            Bin bin = new Bin(initialCapacity);
            bin.addObject(weight);
            addNewBin(bin);
        } else if(node.right.maxRemainingCapacity < weight){
            node = addFirstRec(initialCapacity, weight, node.right);
        } else if(node.left.maxRemainingCapacity >= weight){
            node.left.bin.addObject(weight);
        } else if(node.maxRemainingCapacity >= weight) {
            node.bin.addObject(weight);
        } else {
            node = addFirstRec(initialCapacity, weight, node.right);
        }

        return node;
    }

    public void addAll(int initialCapacity, int[] weights) {
        // todo
    }

    public List<Bin> toList() {
        // todo
        return null;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        stringBuild(sb, root);
        sb.append(")");
        return sb.toString();
    }

    private static void stringBuild(StringBuilder sb, Node node) {
        if(node==null)
            sb.append("null");
        else {
            sb.append(node.getClass().getSimpleName());
            sb.append("(");
            sb.append(node.bin);
            sb.append(", ");
            sb.append(node.height);
            sb.append(", ");
            sb.append(node.maxRemainingCapacity);
            sb.append(", ");
            stringBuild(sb, node.left);
            sb.append(", ");
            stringBuild(sb, node.right);
            sb.append(")");
        }
    }
}

class LinearBinPacking {
    public static List<Bin> linearBinPacking(int initialCapacity, List<Integer> weights) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }
	
	public static Iterable<Integer> allWeights(Iterable<Bin> bins) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;		
	}
}

File: Exámenes/2020 Septiembre/Haskell 2020/AVL.hs
{------------------------------------------------------------------------------
 - Student's name:
 -
 - Student's group:
 -----------------------------------------------------------------------------}

module AVL 
  ( 
    Weight
  , Capacity
  , AVL (..)
  , Bin
  , emptyBin
  , remainingCapacity
  , addObject
  , maxRemainingCapacity
  , height
  , nodeWithHeight
  , node
  , rotateLeft
  , addNewBin
  , addFirst
  , addAll
  , toList
  , linearBinPacking
  , seqToList
  , addAllFold
  ) where

type Capacity = Int
type Weight= Int

data Bin = B Capacity [Weight] 

data AVL = Empty | Node Bin Int Capacity AVL AVL deriving Show


emptyBin :: Capacity -> Bin
emptyBin c
  | c < 0 = error "La capacidad no puede ser negativa"
  | otherwise = B c []

remainingCapacity :: Bin -> Capacity
remainingCapacity (B c _) = c

addObject :: Weight -> Bin -> Bin
addObject w (B c xs)
  | w > c = error "el objeto no cabe dentro del cubo"
  | otherwise = B (c-w) (xs ++ [w])

maxRemainingCapacity :: AVL -> Capacity
maxRemainingCapacity Empty = 0
maxRemainingCapacity (Node b h c lt rt) = c


height :: AVL -> Int
height Empty = 0
height (Node b h c lt rt) = h 


 
nodeWithHeight :: Bin -> Int -> AVL -> AVL -> AVL
nodeWithHeight b h lt rt = Node b h (1+(max (maxRemainingCapacity lt) (maxRemainingCapacity rt))) lt rt


node :: Bin -> AVL -> AVL -> AVL
node b lt rt = Node b (max (height lt) (height rt)) (max (maxRemainingCapacity lt) (maxRemainingCapacity rt)) lt rt 

rotateLeft :: Bin -> AVL -> AVL -> AVL
rotateLeft c l (Node x h2 c2 r1 r2) = node x (node c l r1) r2

addNewBin :: Bin -> AVL -> AVL
addNewBin b Empty = node b Empty Empty
addNewBin b (Node x h c lt rt) 
  | (height rt) - (height lt) > 1 = rotateLeft x lt (addNewBin b rt)
  | otherwise = node x lt (addNewBin b rt)
 
addFirst :: Capacity -> Weight -> AVL -> AVL
addFirst c w Empty = (addNewBin (B (c-w) [w]) Empty)
addFirst c w nod@(Node x h c2 lt rt)
  |w > maxRemainingCapacity nod = addNewBin (B (c-w) [w]) nod
  |w <= maxRemainingCapacity lt = node x (addFirst c w lt) rt
  |w <=remainingCapacity x = node (addObject w x) lt rt
  |otherwise = node x lt (addFirst c w rt)

addAll:: Capacity -> [Weight] -> AVL
addAll c xs = aux c xs Empty
  where
    aux c [] sol = sol
    aux c (x:xs) sol = aux c xs (addFirst c x sol) 

toList :: AVL -> [Bin]
toList Empty = []
toList (Node b h c lt rt) = toList lt ++ [b] ++ toList rt 

{-
	SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
 -}

data Sequence = SEmpty | SNode Bin Sequence deriving Show  

linearBinPacking:: Capacity -> [Weight] -> Sequence
linearBinPacking _ _ = undefined

seqToList:: Sequence -> [Bin]
seqToList _ = undefined

addAllFold:: [Weight] -> Capacity -> AVL 
addAllFold _ _ = undefined



{- No modificar. Do not edit -}

objects :: Bin -> [Weight]
objects (B _ os) = reverse os

  
instance Show Bin where
  show b@(B c os) = "Bin("++show c++","++show (objects b)++")"